### PR TITLE
refactor: Remove ::Params() global from CChainState

### PR DESCRIPTION
**Pull request by maflcko**
The `::Params()` global is verbose and confusing. Also it makes tests a bit harder to write because they'd have to mock a global.

Fix all issues by simply using a member variable that points to the right params.

(Can be reviewed with `--word-diff-regex=.`)

### CODE REVIEW

**Comment by jamesob** [comment]
Concept ACK, will review soon

**Comment by practicalswift** [comment]
Concept ACK

**Comment by fanquake** [comment]
Concept ACK

File: src/validation.h [hunk 1]
```diff
@@ -564,6 +564,8 @@ class CChainState
     //! mempool that is kept in sync with the chain
     CTxMemPool& m_mempool;
 
+    const CChainParams& m_params;
```
Comment (kiminuo) [inline src/validation.h:567]: nit: This member variable looks like the only one without a Doxygen comment in this section. Maybe add one?

Is `protected` better here than `private`? It compiled for me even when the variable is `private`.

**Comment by kiminuo** [comment]
Concept ACK.

It looks like a very nice simplification of the code which makes it harder to make a mistake.

File: src/validation.h (hunk 1 above)
Comment (MarcoFalke) [reply src/validation.h:567]: Thanks, fixed. Also fixed up some typos.

Comment (MarcoFalke) [reply src/validation.h:567]: Actually, changing to `private` didn't compile, so I left this as-is for now.

Comment (kiminuo) [reply src/validation.h:567]: I moved only `const CChainParams& m_params;` to `private` section and it did compile for me. It confuses me but ok.

Comment (MarcoFalke) [reply src/validation.h:567]: `private` and `protected` are the same anyway, unless the class is derived

**Comment by MarcoFalke** [comment]
Rebased

File: src/chain.h [hunk 2]
```diff
@@ -182,7 +182,7 @@ class CBlockIndex
     //!
     //! Note: this value is modified to show BLOCK_OPT_WITNESS during UTXO snapshot
     //! load to avoid the block index being spuriously rewound.
-    //! @sa RewindBlockIndex
+    //! @sa NeedsRedownload
```
Comment (kiminuo) [inline src/chain.h:185]: For the sake of completeness, context for the change is [here](https://github.com/bitcoin/bitcoin/pull/21816).

**Comment by kiminuo** [comment]
The main change in the first commit (fa216cc7) seems to be: "refactor: Add `CChainState` member to `CChainState` class" rather than "refactor: Remove ::Params() global from CChainState".

Also, the following change in the first commit (fa216cc7) https://github.com/bitcoin/bitcoin/pull/21789/commits/fa216cc79267ba25fa3709982626def61f4abe2a#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R2284 can probably be moved to the second commit (fa06aaacf3b9c35d9e30c883bcd7780dcf0bceda).

Changes in fa216cc79267ba25fa3709982626def61f4abe2a & fa06aaacf3b9c35d9e30c883bcd7780dcf0bceda LGTM.

**Comment by jnewbery** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
I think the two commits are nicely split up. The first one removes it from *inside* the functions, the second removes the *arg* of the functions.

File: src/test/util/setup_common.cpp [hunk 3]
```diff
@@ -185,12 +185,12 @@ TestingSetup::TestingSetup(const std::string& chainName, const std::vector<const
     assert(!::ChainstateActive().CanFlushToDisk());
     ::ChainstateActive().InitCoinsCache(1 << 23);
     assert(::ChainstateActive().CanFlushToDisk());
-    if (!::ChainstateActive().LoadGenesisBlock(chainparams)) {
+    if (!::ChainstateActive().LoadGenesisBlock()) {
```
Comment (jnewbery) [inline src/test/util/setup_common.cpp:188]: If you touch this branch again, consider removing the local `chainparams` reference and just calling `Params()` in the call to `PeerManager::make()` below.

**Comment by jnewbery** [comment]
ACK 2222da832066d06be3ff8752e49008fef71eb668

File: src/test/util/setup_common.cpp (hunk 3 above)
Comment (MarcoFalke) [reply src/test/util/setup_common.cpp]: The line will be touched again in the follow-up that removes `::Params()`, so I'll leave it as is for now.

**Comment by MarcoFalke** [comment]
Rebased

**Comment by jnewbery** [comment]
ACK fa0d9211ef87a682573aaae932c0c440acbcb8a8

**Comment by kiminuo** [comment]
utACK fa0d9211

File: src/test/interfaces_tests.cpp [hunk 4]
```diff
@@ -98,7 +98,7 @@ BOOST_AUTO_TEST_CASE(findCommonAncestor)
     auto* orig_tip = active.Tip();
     for (int i = 0; i < 10; ++i) {
         BlockValidationState state;
-        m_node.chainman->ActiveChainstate().InvalidateBlock(state, Params(), active.Tip());
+        m_node.chainman->ActiveChainstate().InvalidateBlock(state, active.Tip());
```
Comment (kiminuo) [inline src/test/interfaces_tests.cpp:101]: It seems to me that `#include <chainparams.h>` is no longer needed.

Comment (MarcoFalke) [reply src/test/interfaces_tests.cpp:101]: Thanks, will try to remove if I have to force push

**Review by theStack — APPROVED** [review]
ACK fa0d9211ef87a682573aaae932c0c440acbcb8a8 🍉 
Nice refactoring idea, glad to see this simplification of the CChainState interface.

**Pull request merged**